Add temperature to the light color mode parameter fallbacks

nmaggioni opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Some warm/cold white lights only understand the `color_temp` parameter, and as highlighted in #45111 the [color profiles](https://github.com/home-assistant/core/blob/3bb9be23820c94dba75db1001878f06c81800db6/homeassistant/components/light/light_profiles.csv) only send XY values.

This PR indirectly solves that issue by adding an additional color temperature fallback mode when computing alternative color formats that are supported by the device: if it must resort to the color temperature attribute, it is now computed from the original XY values instead of ignored.

This incidentally enables devices like the IKEA Tradfri white-only bulbs to be controlled by specifying any color format and having it automatically converted instead of needing to manually specify a color temperature.

#### Future-proofing?

Abiding to #79591 the main attribute being set is `ATTR_COLOR_TEMP_KELVIN`, but given that some devices still need `ATTR_COLOR_TEMP` and the conversion would happen _before_ the rest of the color mode fallback code runs ([`preprocess_turn_on_alternatives`](https://github.com/home-assistant/core/blob/6a89b3a135ebaff1f3c8f6dbc8e32974241b5e8f/homeassistant/components/light/__init__.py#L295) runs [before](https://github.com/home-assistant/core/blob/6a89b3a135ebaff1f3c8f6dbc8e32974241b5e8f/homeassistant/components/light/__init__.py#L430) the actual [fallback conversions](https://github.com/home-assistant/core/blob/6a89b3a135ebaff1f3c8f6dbc8e32974241b5e8f/homeassistant/components/light/__init__.py#L470)), at the moment I've resorted to manually deriving and applying the `ATTR_COLOR_TEMP` attribute (see commit [007e863](https://github.com/home-assistant/core/commit/007e86384a62e5def1a2b91cf93ee6f276ad7450)) once again.

This approach would rely on manually reaping any occurrence of `ATTR_COLOR_TEMP` when getting rid of it in the future, so I don't particularly like it. What could be a cleaner solution? Is there even a real need for it or is the one implemented here good enough already? Calling `preprocess_turn_on_alternatives` after the conversion would be counterintuitive but something of that sort would be needed in order to re-parse the Kelvin value and automatically fill in the legacy Mired one.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #45111 
- The XY to Temperature conversion code uses [McCamy's approximation](https://onlinelibrary.wiley.com/doi/10.1002/col.5080170211) of the CCT curve, considered to be reliable for ranges between 2000 K and 10_000 K, thus enough for this use case. More accurate but also more complex models exist.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- ~~Documentation added/updated for [www.home-assistant.io][docs-repository]~~

If the code communicates with devices, web services, or third-party tools:

- ~~The [manifest file][manifest-docs] has all fields filled out correctly.~~  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- ~~New or updated dependencies have been added to `requirements_all.txt`.~~  
      Updated by running `python3 -m script.gen_requirements_all`.
- ~~For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.~~
- ~~Untested files have been added to `.coveragerc`.~~

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

nmaggioni committed:
Add color temperature to the color mode fallbacks

nmaggioni committed:
Manually add ATTR_COLOR_TEMP since ATTR_COLOR_TEMP_KELVIN is pre-parsed

nmaggioni committed:
Include the legacy ATTR_COLOR_TEMP attribute in the tests

nmaggioni commented:
Still valid IMO, and branch updated to solve recent conflicts.
Respectfully pinging @frenck due to its recent work on these parts of the lighting system.

ClayBenson94 commented:
Given the existence of the color util functions, and the layout of the fallbacks, this change seems totally natural to me (just adding another color mode support in an existing function)

Happy to try and pull this down locally and try it out myself - I am using Hue ambiance bulbs, which don't seem to react to XY colors being sent in `light.turn_on`

nmaggioni commented:
Ping.

nmaggioni committed:
Merge branch 'dev' into nm_lights_color_temperature_fallback

emontnemery commented:
@nmaggioni I don't understand the reason for this change. Why would anyone want to set a light which only supports white to a color; the light will anyway just be white? If the real driver is to improve light profiles, I'm not so sure; light profiles are very long overdue for an overhaul; we should either remove light profiles or modernize them with UI support and so on.

nmaggioni commented:
> I don't understand the reason for this change. Why would anyone want to set a light which only supports white to a color;

The linked issue should contain sufficient justification :)
As of today, applying a `profile` to bulbs that only support `color_temperature` simply changes their brightness. 

> the light will anyway just be white and it will be very tricky for the user to work out which part of the color space can be rendered by the light?

This PR only adds internal color conversion utilities to automatically translate a requested state into one that the final device will actually understand and act on: the UI certainly won't expose full color controls, so if a user wants to go in there and manually tinker with the services' payloads nobody will be surprised when the bulb won't react to "impossible" states (just like it does without these changes, honestly). Also you can't say that the actual behavior of requesting a color/temperature change and seeing only a brightness response is any better.

The only actual behavioral change this code brings is that HA's default color presets will work as expected in terms of temperature, so users of temperature-only bulbs won't need to implement custom profiles and/or dedicated automations to replicate the states that are built into the stock apps for, let's say, my IKEA Tradfri bulbs that come with three temperature settings that can be toggled by the stock remote.

Since [you contributed code related to this exact matter](https://github.com/home-assistant/core/commit/5f2326fb5756b85255cec27d3b2172c44bc66058) I'm surprised by your criticism, as this PR is basically just an extra `elif` that wasn't implemented at the time due to, I guess, simply not having encountered these somewhat weird bulbs.

> If the real driver is to improve light profiles, light profiles should be improved instead. Light profiles are very long overdue for an overhaul; we should either remove light profiles altogether and rely on scenes or modernize them with UI support and so on.

I'm totally with you on this one, but back in January this was the biggest contribution I could make and it worked pretty well for me for quite a long time. I don't really care if this gets merged or profiles get improved: the goal here is to make light profiles work with all bulbs. The difference between the two alternatives is that this PR will immediately fix the issue, while I'm not aware of any ongoing effort to improve or replace light profiles in general.

emontnemery commented:
> Since you contributed code related to this exact matter I'm surprised by your criticism

It's not criticism, I want to understand how this PR is useful.

> this PR is basically just an extra elif that wasn't implemented at the time due to, I guess, simply not having encountered these somewhat weird bulbs

That's not the case, a light which supports colors can often approximate a white with a given color temperature with somewhat acceptable accuracy, but a light which supports white with adjustable color temperature will not be able to render any given color. Such a lights can only render a very very small part of the color space supported by a color light; the line in the xy-color space illustration below shows which of the colors in the xy-color space can be expressed as white with a given color temperatures: 
![image](https://github.com/home-assistant/core/assets/14281572/bc92b1ad-1f42-4c14-9623-bff2b027f878)

Bulbs with adjustable color temperatures are not at all weird, I'd say they are very common.

> The only actual behavioral change this code brings is that HA's default color presets will work as expected in terms of temperature, so users of temperature-only bulbs won't need to implement custom profiles and/or dedicated automations to replicate the states that are built into the stock apps for, let's say, my IKEA Tradfri bulbs that come with three temperature settings that can be toggled by the stock remote.

You're losing me here. I thought the rationale for this PR is precisely to allow users to add their own profiles and use those profiles  for white lights with adjustable color temperature? The [example profiles which come with Home Assistant](https://github.com/home-assistant/core/blob/master/homeassistant/components/light/light_profiles.csv) are not suitable, because they will set the lights to various, mostly red-ish, colors, and do not, at least not intentionally, replicate the presets of IKEA trådfri remotes.

Again, light profiles are very long overdue for a major overhaul, scenes are probably what you should use now.

emontnemery reviewed (changes requested):
@nmaggioni I discussed your PR with @MartinHjelmare. While we still don't agree with the proposed use case (allow controlling white lights with adjustable color temperatures from color profiles) we think adding the service call conversion is OK.

Just a minor comment to please add a source for the cubic approximation when converting from x/y-color to a color temperature.

nmaggioni committed:
Apply suggestions from code review

Co-authored-by: Erik Montnemery <erik@montnemery.com>

nmaggioni commented on homeassistant/util/color.py:
```diff
@@ -576,6 +576,17 @@ def _white_levels_to_color_temperature(
+
+    Uses McCamy's approximation, close enough for uses between 2000 K and 10_000 K.
+    """
+    n = (x - 0.3320) / (0.1858 - y)
+    CCT = 437 * (n**3) + 3601 * (n**2) + 6861 * n + 5517
```
My sources at the time were indeed the original paper and that STMicro's application note, but now my access to Wiley's library has expired so I can't check the fulltext. The [erratum](https://onlinelibrary.wiley.com/doi/epdf/10.1002/col.5080180222) is publicly available though, and the constants roughly match - excluding the opposite signs which probably refer to some other step of the calculation.

I've tested all the three coefficients combinations I could find on the web (citing publications, StackOverflow, application notes) and they all give sufficiently similar results (or at least differ in amounts that aren't visible to the naked eye when rendered by a lightbulb), so I think it's pretty safe to link to the original paper.

Linking to STMicro's notes would be handy but I have no trust at all in that link staying up indefinitely, while at least the paper's DOI URL is pretty much guarateed to always resolve.

nmaggioni replied:
I've just added the link to the application note to the commit description though, so it's not completely lost in time.

nmaggioni committed:
Add citation for McCamy's approximation formula

If still existing, also see page 3 of https://www.st.com/resource/en/application_note/an5638-how-correlated-color-temperature-is-calculated-by-vd6283-stmicroelectronics.pdf

emontnemery committed:
Update homeassistant/util/color.py

emontnemery reviewed (approved):
Thanks, @nmaggioni 👍
